packages/react/src/merge-props/mergeProps.ts

```diff
@@ -29,110 +32,142 @@ type MergableProps<T extends React.ElementType> =
  * @param props props to merge.
  * @returns the merged props.
  */
-export function mergeProps<Tag extends React.ElementType>(
-  ...props: MergableProps<Tag>[]
-): WithBaseUIEvent<React.ComponentPropsWithRef<Tag>> {
-  if (props.length === 0) {
-    return {} as WithBaseUIEvent<React.ComponentPropsWithRef<Tag>>;
+/* eslint-disable id-denylist */
+export function mergeProps<T extends ElementType>(a: InputProps<T>, b: InputProps<T>): PropsOf<T>;
+export function mergeProps<T extends ElementType>(a: InputProps<T>, b: InputProps<T>): PropsOf<T>;
+export function mergeProps<T extends ElementType>(
+  a: InputProps<T>,
+  b: InputProps<T>,
+  c: InputProps<T>,
+): PropsOf<T>;
+export function mergeProps<T extends ElementType>(
+  a: InputProps<T>,
+  b: InputProps<T>,
+  c: InputProps<T>,
+  d: InputProps<T>,
+): PropsOf<T>;
+export function mergeProps<T extends ElementType>(
+  a: InputProps<T>,
+  b: InputProps<T>,
+  c: InputProps<T>,
+  d: InputProps<T>,
+  e: InputProps<T>,
+): PropsOf<T>;
+export function mergeProps(a: any, b: any, c?: any, d?: any, e?: any) {
```

Comment:
Although 6+ sets of props is unlikely, this technically allowed an infinite number of objects to merge before, so it'd be a breaking change?

Reply:
Yes, but isn't base-ui still in alpha? If it really is a problem, we could export `mergePropsN` as `mergeProps`.

Reply:
Yeah, we just need to mark them as `breaking change`, which is okay still. However I think such a utility should probably be variadic for the public export, so `mergePropsN` should be good.

Reply:
I think we shouldn't expose the slower variadic version by default. There's not a single case inside the codebase where we need to merge 6+ set of props, I don't think it's more likely to happen outside the codebase. If it does, it's nice that users have a moment to stop and think about what they're doing.